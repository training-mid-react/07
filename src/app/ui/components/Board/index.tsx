import { useState } from 'react';
import { BoardMatrix, Player } from '../../../core/interfaces/board-state';
import './style.scss';

interface Props{
  board: BoardMatrix,
  handleSelectColumn: (column: number) => void
  currentPlayer: Exclude<Player, 'draw'>
}
export const Board = ({board, handleSelectColumn, currentPlayer}: Props) => {
  
  const [animatedCell, setAnimatedCell] = useState<{ col: number; row: number } | null>(null);

  const handleCellClick = (colIndex: number) => {
    const rowIndex = board.findIndex(row => row[colIndex] === null);
    if (rowIndex !== -1) {
      setAnimatedCell({ col: colIndex, row: rowIndex });
      setTimeout(() => {
        handleSelectColumn(colIndex);
        setAnimatedCell(null);
      }, 500); 
    }
  };

  return (
    
    <div className="board">
      {
        board.map((row, rowIndex) => {
          return (
            <div 
              key={`row_${rowIndex}`} 
              className='board__row'
            >
              {row.map((cell, colIndex) => {
                const isFalling = animatedCell?.col === colIndex && animatedCell.row >= rowIndex;
                const fallDistance = isFalling ? `${50 * (animatedCell!.row - rowIndex)}px` : '0px';
                const animationFall = {
                  transform: `translateY(${fallDistance})`,
                };
                return( 
                  <div 
                    key={`cell_${colIndex}`} 
                    data-testid={`cell_${rowIndex}_${colIndex}`}
                    className={`board__col ${cell || ''} ${isFalling ? `falling--${currentPlayer}` : ''}`}  
                    onClick={() => handleCellClick(colIndex)}
                    style={isFalling ? animationFall : {}}
                  />
                );
              })}
            </div>
          );
        })
      }
      
    </div>
  );
};
