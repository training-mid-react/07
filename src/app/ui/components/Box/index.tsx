import { CELL } from "@/app/core/interfaces/game.interface";
import "./style.scss"

interface Props {
    value: CELL;
    onClick: () => void;
}

const Box = ({ value, onClick }: Props) => {
    return (
        <div data-testid="box" className="box" onClick={onClick}>
            {
                value && <div style={{backgroundColor: value.color}} className="box__chip"></div>
            }
        </div>
    );
}

export default Box;