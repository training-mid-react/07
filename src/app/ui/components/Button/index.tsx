import React from 'react';
import './style.scss';

interface ButtonProps {
  children: React.ReactNode;
  className?: string;
  onClick: () => void;
  disabled?: boolean;
}

const Button: React.FC<ButtonProps> = ({ children, onClick, disabled, className }) => {
  return (
    <button data-testid='button-component' className={className || 'custom-button'} onClick={onClick} disabled={disabled}>
      {children}
    </button>
  );
};

export default Button;
