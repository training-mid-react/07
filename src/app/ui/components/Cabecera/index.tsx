
import "./style.scss";

interface Props {
    currentPlayer: number
}

const Cabecera = ({ currentPlayer }: Props) => {
    return (
        <>
            <section className="cabecera">
                <div>
                    <h2>Jugador 1</h2>
                    <div className="ficha-cabecera ficha-cabecera--jugador1"></div>
                </div>
                <div>
                    <h2>Jugador 2</h2>
                    <div className="ficha-cabecera ficha-cabecera--jugador2"></div>
                </div>
            </section>
            <h3>Turno del jugador: {currentPlayer}</h3>
        </>

    )
}

export default Cabecera 