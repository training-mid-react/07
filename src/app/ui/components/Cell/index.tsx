import { FC } from 'react';
import { PLAYER } from '../../../../core/interfaces';
import './style.scss';

export interface CellProps {
  value: PLAYER | null;
}

export const Cell: FC<CellProps> = ({ value }) => {

  return (
    <>
    <div className={`cell cell__${value}`} />
    <div className='cell__frame'></div>
    </>
  );
};
