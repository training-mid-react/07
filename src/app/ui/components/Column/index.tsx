import './style.scss';
import { useDropCell } from '../../../core/hooks/useDropCell';

interface ColumnProps {
  columnIndex: number;
  dropPiece: (column: number) => void;
  columnData: number[];
}

function Column({ columnIndex, dropPiece, columnData }: ColumnProps) {
  const { handleColumnClick, droppingCell } = useDropCell({ columnIndex, dropPiece, columnData })
  return (
    <div className="column" onClick={handleColumnClick}>
      {columnData.map((cell, rowIndex) => {
        const isDropping = droppingCell === rowIndex;
        const cellClass = `cell ${cell === 1 ? 'cell--player1' : cell === 2 ? 'cell--player2' : ''} ${isDropping ? 'cell--dropping' : ''}`;
        return (
          <div
            key={rowIndex}
            className={cellClass}
          />
        );
      })}
    </div>
  );
};

export default Column;
