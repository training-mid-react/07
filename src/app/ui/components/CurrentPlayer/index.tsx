import './style.scss';

interface ICurrentPlayerProps {
  currentPlayer: number;
}

export const CurrentPlayer = ({ currentPlayer }: ICurrentPlayerProps) => {
  return (
    <div className="current-player">
      Es el turno del Jugador: 
      <span className={currentPlayer === 1 ? 'current-player__player1' : 'current-player__player2'}>
        {currentPlayer}
      </span>
    </div>
  );
};
