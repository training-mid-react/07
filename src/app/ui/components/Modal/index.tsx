import { ReactNode } from 'react';
import './style.scss';

interface Props {
    titleModal:string;
    labelButton: string;
    buttonAction: () => void;
    children: ReactNode | ReactNode[];
}
export const Modal = ({titleModal,labelButton,buttonAction, children}: Props) => {
  return (
    <article className="modal">
      <div className='modal__wrapper'>
        <h2 className='modal__title'>{titleModal}</h2>
        <section className="modal__content">
          {children}
        </section>

        <section className='modal__action'>
          <button
            className='modal__button'
            type='button' 
            onClick={buttonAction}
          >
            {labelButton}
          </button>
        </section>
      </div>
    </article>
  );
};
