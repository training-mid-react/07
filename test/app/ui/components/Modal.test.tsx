import React from 'react';
import { vi } from 'vitest';
import { fireEvent, render } from '@testing-library/react';

import {Modal} from '../../../../src/app/ui/components/Modal';

describe('Modal', () => {

  // ===> Initial state
  const titleModal = 'titleModal';
  const labelButton = 'buttonLabel';

  test('Matches the snapshot', () => {
    const buttonAction = vi.fn();

    const {asFragment} = render(
      <Modal
        titleModal={titleModal}
        labelButton={labelButton}
        buttonAction={buttonAction}
      >
         Content Modal
      </Modal>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Expect to render the modal', () => {
    const buttonAction = vi.fn();

    const {getByText} = render(
      <Modal
        titleModal={titleModal}
        labelButton={labelButton}
        buttonAction={buttonAction}
      >
         Content Modal
      </Modal>
    );

    expect(getByText(titleModal)).toBeDefined();
    expect(getByText(labelButton)).toBeDefined();
  });

  test('Expect to call the button action', () => {
    const buttonAction = vi.fn();

    const {getByText} = render(
      <Modal
        titleModal={titleModal}
        labelButton={labelButton}
        buttonAction={buttonAction}
      >
         Content Modal
      </Modal>
    );

    fireEvent.click(getByText(labelButton));
    expect(buttonAction).toHaveBeenCalledTimes(1);
  });
});