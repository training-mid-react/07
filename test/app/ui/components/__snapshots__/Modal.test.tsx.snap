// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > Matches the snapshot 1`] = `
<DocumentFragment>
  <article
    class="modal"
  >
    <div
      class="modal__wrapper"
    >
      <h2
        class="modal__title"
      >
        titleModal
      </h2>
      <section
        class="modal__content"
      >
        Content Modal
      </section>
      <section
        class="modal__action"
      >
        <button
          class="modal__button"
          type="button"
        >
          buttonLabel
        </button>
      </section>
    </div>
  </article>
</DocumentFragment>
`;
